src/graphics/particles.js

```diff
@@ -146,223 +160,232 @@ Crafty.c("Particles", {
-                    if (this.jitter) {
-                        currentParticle.position.x += this.jitter * this.RANDM1TO1();
-                        currentParticle.position.y += this.jitter * this.RANDM1TO1();
+                if (particle.timeToLive > 0) {
+
+                    // Calculate the new position based on gravity
+                    particle.directionX += gravityX;
+                    particle.directionY += gravityY;
+                    particle.positionX += particle.directionX;
+                    particle.positionY += particle.directionY;
+                    if (jitter) {
+                        particle.positionX += jitter * RANDM1TO1();
+                        particle.positionY += jitter * RANDM1TO1();
                     }
-                    currentParticle.timeToLive--;
 
                     // Update colours
-                    var r = currentParticle.colour[0] += currentParticle.deltaColour[0];
-                    var g = currentParticle.colour[1] += currentParticle.deltaColour[1];
-                    var b = currentParticle.colour[2] += currentParticle.deltaColour[2];
-                    var a = currentParticle.colour[3] += currentParticle.deltaColour[3];
-
-                    // Calculate the rgba string to draw.
-                    draw = [];
-                    draw.push("rgba(" + (r > 255 ? 255 : r < 0 ? 0 : ~~r));
-                    draw.push(g > 255 ? 255 : g < 0 ? 0 : ~~g);
-                    draw.push(b > 255 ? 255 : b < 0 ? 0 : ~~b);
-                    draw.push((a > 1 ? 1 : a < 0 ? 0 : a.toFixed(2)) + ")");
-                    currentParticle.drawColour = draw.join(",");
-
-                    if (!this.fastMode) {
-                        draw[3] = "0)";
-                        currentParticle.drawColourEnd = draw.join(",");
-                    }
-
-                    this.particleIndex++;
-                } else {
-                    // Replace particle with the last active
-                    if (this.particleIndex !== this.particleCount - 1) {
-                        this.particles[this.particleIndex] = this.particles[this.particleCount - 1];
-                    }
-                    this.particleCount--;
+                    c = particle.colourR + particle.deltaColourR;
+                    particle.colourR = c > 255 ? 255 : c < 0 ? 0 : ~~c;
+                    c = particle.colourG + particle.deltaColourG;
+                    particle.colourG = c > 255 ? 255 : c < 0 ? 0 : ~~c;
+                    c = particle.colourB + particle.deltaColourB;
+                    particle.colourB = c > 255 ? 255 : c < 0 ? 0 : ~~c;
+                    c = particle.colourA + particle.deltaColourA;
+                    particle.colourA = c > 1 ? 1 : c < 0 ? 0 : (~~(c * 100)) / 100;
+
+                    // Decrease particle's lifespan
+                    particle.timeToLive--;
```

Comment:
With the current implementation, this is really `framesToLive` -- at some point it should get updated to use the proper `dt`.  That would maybe involve changing how velocity is defined.